### ISSUE TITLE
Support for Adaptive Instance Normalization

### ISSUE THREAD

**Issue by christinazavou** [NVIDIA/MinkowskiEngine#345]
Hello,

I need to implement an [Adaptive Instance Normalization](https://paperswithcode.com/method/adaptive-instance-normalization) layer introduced in [Arbitrary Style Transfer in Real-time with Adaptive Instance Normalization](https://arxiv.org/abs/1703.06868).

From my understanding, such a layer has zero learnable parameters.

![image](https://user-images.githubusercontent.com/15656466/115410792-9bf01d80-a1fb-11eb-8345-8ab61489edd6.png)

What is needed, is given a content feature tensor (x) and a style feature tensor (y) to calculate the mean and variance per channel and per instance both for x (μ(x), σ(x)) and y (μ(y), σ(y)), and then use substraction, addition, multiplication.

I first thought of getting the features per instance and channel using 
```x.coords_man.get_row_indices_per_batch(x.coords_key)```,
but this returns a list of lists, where these lists -indicating the coordinate indices per batch instance- have different lengths... therefore, I wouldn't know how to proceed in finding the mean and variance per instance in a "torch-with-gradients" way !

Then I realized there exists an implementation for the Instance Normalization layer (which has learnable parameters - a weight and a bias): ```MinkowskiInstanceNorm```.

I though I could follow the implementation of that layer, and remove any code related to backward pass and ctx... therefore I now have this code:

```
from torch.nn import Module
from MinkowskiEngine.Common import GlobalPoolingMode, get_minkowski_function
from MinkowskiEngine import SparseTensor
from MinkowskiEngine.MinkowskiCoords import CoordsKey
from MinkowskiEngine.MinkowskiBroadcast import OperationType, operation_type_to_int


def minkowski_adain_function(content_feat,
                             style_feat,
                             mode=GlobalPoolingMode.AUTO,
                             in_coords_key=None,
                             glob_coords_key=None,
                             coords_manager=None):
    assert isinstance(mode, GlobalPoolingMode), \
        f"Mode must be an instance of GlobalPoolingMode, {mode}"
    if glob_coords_key is None:
        glob_coords_key = CoordsKey(in_coords_key.D)

    gpool_forward = get_minkowski_function('GlobalPoolingForward', content_feat)
    broadcast_forward = get_minkowski_function('BroadcastForward', content_feat)
    add = operation_type_to_int(OperationType.ADDITION)
    multiply = operation_type_to_int(OperationType.MULTIPLICATION)

    mean = content_feat.new()
    num_nonzero = content_feat.new()

    cpp_in_coords_key = in_coords_key.CPPCoordsKey
    cpp_glob_coords_key = glob_coords_key.CPPCoordsKey
    cpp_coords_manager = coords_manager.CPPCoordsManager

    mean, num_nonzero = gpool_forward(content_feat, cpp_in_coords_key,
                                      cpp_glob_coords_key,
                                      cpp_coords_manager, True, mode.value)
    # X - \mu
    centered_feat = broadcast_forward(content_feat, -mean, add,
                                      cpp_in_coords_key,
                                      cpp_glob_coords_key,
                                      cpp_coords_manager)

    # Variance = 1/N \sum (X - \mu) ** 2
    variance, num_nonzero = gpool_forward(centered_feat**2,
                                          cpp_in_coords_key,
                                          cpp_glob_coords_key,
                                          cpp_coords_manager, True,
                                          mode.value)

    # norm_feat = (X - \mu) / \sigma
    inv_std = 1 / (variance + 1e-8).sqrt()
    norm_feat = broadcast_forward(centered_feat, inv_std, multiply,
                                  cpp_in_coords_key, cpp_glob_coords_key,
                                  cpp_coords_manager)
    return norm_feat


class MinkowskiAdaIN(Module):
    r"""A instance normalization layer for a sparse tensor.

    """

    def __init__(self, mode=GlobalPoolingMode.AUTO):
        r"""
        Args:

            mode (GlobalPoolingModel, optional): The internal global pooling computation mode.
        """
        Module.__init__(self)
        self.mode = mode

    def __repr__(self):
        return self.__class__.__name__

    def forward(self, content, style):
        assert isinstance(content, SparseTensor)
        assert isinstance(style, SparseTensor)

        content_norm = minkowski_adain_function(content.F, 
                                                style.F,
                                                self.mode, 
                                                input.coords_key,
                                                None, 
                                                input.coords_man)

        return SparseTensor(
            content_norm,
            coords_key=input.coords_key,
            coords_manager=input.coords_man)
```

However, this is currently dummy, as the calculations needed inside ```minkowski_adain_function``` are:
```
style_mean, style_std = ...
content_mean, content_std =...
norm_feat = ((content_feat - content_mean) / content_std) * style_std + style_mean
```

Could you please assist me on how to implement this function as I don't know what is happening behing ```gpool_forward``` and ```broadcast_forward```?

p.s. I've tagged this as a "feature request" as I believe AdaIN is currently a commonly used layer in the area of Deep Learning.

Thanks,

------

on a side note,
I wanted to see the relation of torch InstanceNorm1d and MinkowskiInstanceNorm, as i haven't used either before, and got this
````
import torch
from torch.nn.modules.instancenorm import InstanceNorm1d
import numpy as np


input_np = np.array([
    [[1., 2, 3, 4]],
    [[1, 1, 1, 1]]
])

input = torch.from_numpy(input_np)

m = InstanceNorm1d(4)
output = m(input)
print(output)

#tensor([[[-1.3416, -0.4472,  0.4472,  1.3416]],
#      [[ 0.0000,  0.0000,  0.0000,  0.0000]]], dtype=torch.float64)

from MinkowskiEngine import SparseTensor
coords = torch.from_numpy(np.random.random((2, 3))*100)
sinput = SparseTensor(input.squeeze(1), coords)
from MinkowskiEngine.MinkowskiNormalization import MinkowskiInstanceNorm
m = MinkowskiInstanceNorm(4)
output = m(sinput)
print(output.F)

# tensor([[0., 0., 0., 0.],
#        [0., 0., 0., 0.]], dtype=torch.float64, grad_fn=<AddBackward0>)
```
any idea what i'm missing here?

thanks! :)

**Comment by chrischoy** [comment]
I think you can replace the $\sigma(y), \mu(y)$ with `MinkowskiGlobalMaxPooling(y)` or `GlobalMeanPooling(y)` to get the weights and bias.
